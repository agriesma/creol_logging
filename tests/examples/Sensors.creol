/*
 * Sensors.creol -- Primitive sensor networks.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Network
begin
  with Radio
    op send (in data: Data)
end

interface Radio
begin
  with Network
    op send (in data: Data)
    op send (in update: Data)
  with Controller
    op write (in data: Data)
    op read (out data: Data)
    op setPower (in power: Int)
    op setState (in state: Int)
    op getChannelStatus (out status: Int)
    op getError (out error: Int)
end

interface Controller
begin
  with Sensor
    op write (in value: Data)
end

interface Sensor
begin
  with Controller
    op setResolution (in res: Data)
    op setEncoding (in encoding: Data)
    op switchOn
    op switchOff
end

interface PassiveSensor inherits Sensor
begin
  with Controller
    op read(out value: Data)
  with Radio
    op setRate (in newRate: Duration)
end

interface ActiveSensor inherits Sensor
begin
  with Controller
    op setRate(in rate: Duration)
end

class PassiveTempSensor contracts PassiveSensor
begin
  with Controller
    op setResolution (in res: Data) == skip
    op setEncoding (in encoding: Data) == skip
    op switchOn == skip
    op switchOff == skip
    op read (out value: Data) == value := 37
  with Radio
    op setRate (in newRate: Duration) == skip
end

class SimpleSensorNode(sensor: PassiveSensor, radio: Radio, size: Int,
		       initialRate: Duration)
  contracts Controller
begin
  var buffer: List[Data]
  var slots: Int
  var t0: Time
  var rate: Duration

  op init == buffer := nil; slots := size; rate := initialRate; t0 := now

  op run == read(;); await now >= t0 + rate; posit now <= t0 + rate; t0 := now;
    !run()

  op read == var temp: Data; sensor.read(; temp); posit now <= t0;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op packAndSend == radio.write(this -| now -| buffer;); buffer := nil;
    slots := size

  with Radio
    op setRate(in newRate: Duration) ==
      if newRate < rate && now > t0 + newRate then t0 := now; read(;) end;
      rate := newRate; posit now < t0 + rate

  with Sensor
    op write(in data: Data) == skip
end

class TwoSimpleSensorNode(sensor0: PassiveSensor, sensor1: PassiveSensor,
		          radio: Radio, size: Int,
		          initialRate0: Duration, initialRate1: Duration)
  contracts Controller
begin
  var buffer: List[Data]
  var slots: Int
  var t0: Time
  var t1: Time
  var rate0: Duration
  var rate1: Duration

  op init ==
    buffer, slots, rate0, rate1, t0, t1 :=
      nil, size, initialRate0, initialRate1, now, now

  op run == !run0() ; !run1()

  op run0 == read0(;); await now >= t0 + rate0; posit now <= t0 + rate0;
    t0 := now; !run0()

  op read0 == var temp: Data; sensor0.read(; temp); posit now <= t0;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op run1 == read1(;); await now >= t1 + rate1; posit now <= t1 + rate1;
    t1 := now; !run1()

  op read1 == var temp: Data; sensor1.read(; temp); posit now <= t1;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op packAndSend == radio.write(this -| now -| buffer;); buffer := nil;
    slots := size

  with Radio
    op setRate(in newRate: Duration) ==
      if newRate < rate0 && now > t0 + newRate then t0 := now; read0(;) end;
      rate0 := newRate; posit now < t0 + rate0

  with Sensor
    op write(in data: Data) == skip
end

class ActiveSensor(controller: Controller, initialRate: Duration)
  contracts Sensor
begin
  var rate: Duration
  var t0: Time

  op init == rate, t0 := initialRate, now

  op run == await now >= t0 + rate; controller.write(37;);
    posit now <= t0 + rate; t0 := now

  with Controller
    op setRate(in newRate: Duration) ==
      if newRate < rate && now > t0 + newRate then
        controller.write(37;); t0 := now
      end;
      rate := newRate; posit now < t0 + rate

    op setResolution(in data: Data) == skip

    op setEncoding(in data: Data) == skip

    op switchOn == skip

    op switchOff == skip
end
