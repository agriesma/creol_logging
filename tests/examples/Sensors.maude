load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "PassiveTempSensor" : Class | Inh: noInh, Param: noVid, Att: noSubst,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "setResolution" : Method | Param: "res", Att: "res" |-> null,
              "_" |-> null, Code: skip ; return ( emp ) >,
            < "setEncoding" : Method | Param: "encoding", Att: "encoding" |->
              null, "_" |-> null, Code: skip ; return ( emp ) >,
            < "switchOn" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
            < "switchOff" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
            < "read" : Method | Param: noVid, Att: "value" |-> null, "_" |->
              null, Code: assign( "value" ; int(37) ) ; return ( "value" ) >,
            < "setRate" : Method | Param: "newRate", Att: "newRate" |-> null,
              "_" |-> null, Code: skip ; return ( emp ) >,
      Ocnt: 0 >
    
    < "SimpleSensorNode" : Class | Inh: noInh, Param: "sensor", "radio",
      "size", "initialRate", Att: "sensor" |-> null, "radio" |-> null, "size"
      |-> null, "initialRate" |-> null, "buffer" |-> null, "slots" |-> null,
      "t0" |-> null, "rate" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "buffer" ; list(emp) ) ;
              assign( "slots" ; "size" ) ; assign( "rate" ; "initialRate" ) ;
              assign( "t0" ; now ) ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null,
              Code: call( "label:0" ; "this" ; "read" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ; await ">=" ( now
              :: "+" ( "t0" :: "rate" ) ) ; posit "<=" ( now :: "+" ( "t0" ::
              "rate" ) ) ; assign( "t0" ; now ) ;
              call( "label:1" ; "this" ; "run" ; emp ) ; free( "label:1" ) ;
              return ( emp ) >,
            < "read" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "temp" |-> null, "_" |-> null,
              Code: call( "label:0" ; "sensor" ; "read" ; emp ) ;
              get( "label:0" ; "temp" ) ; free( "label:0" ) ; posit
              "<=" ( now :: "t0" ) ; await ">" ( "slots" :: int(0) ) ;
              assign( "buffer" ; "|-" ( "buffer" :: "temp" ) ) ;
              assign( "slots" ; "-" ( "slots" :: int(1) ) ) ; if
              "=" ( "slots" :: int(0) ) th
              call( "label:1" ; "this" ; "packAndSend" ; emp ) ;
              get( "label:1" ; noVid ) ; free( "label:1" ) el skip fi ;
              return ( emp ) >,
            < "packAndSend" : Method | Param: noVid, Att: "label:0" |-> null,
              "_" |-> null, Code: call( "label:0" ; "radio" ; "write" ;
              "-|" ( "this" :: "-|" ( now :: "buffer" ) ) ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ;
              assign( "buffer" ; list(emp) ) ; assign( "slots" ; "size" ) ;
              return ( emp ) >,
            < "setRate" : Method | Param: "newRate", Att: "newRate" |-> null,
              "label:0" |-> null, "_" |-> null, Code: if
              "&&" ( "<" ( "newRate" :: "rate" ) :: ">" ( now :: "+" ( "t0"
              :: "newRate" ) ) ) th assign( "t0" ; now ) ;
              call( "label:0" ; "this" ; "read" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) el skip fi ;
              assign( "rate" ; "newRate" ) ; posit "<" ( now :: "+" ( "t0" ::
              "rate" ) ) ; return ( emp ) >,
            < "write" : Method | Param: "data", Att: "data" |-> null, "_" |->
              null, Code: skip ; return ( emp ) >,
      Ocnt: 0 >
    
    < "TwoSimpleSensorNode" : Class | Inh: noInh, Param: "sensor0",
      "sensor1", "radio", "size", "initialRate0", "initialRate1",
      Att: "sensor0" |-> null, "sensor1" |-> null, "radio" |-> null, "size"
      |-> null, "initialRate0" |-> null, "initialRate1" |-> null, "buffer"
      |-> null, "slots" |-> null, "t0" |-> null, "t1" |-> null, "rate0" |->
      null, "rate1" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "buffer", "slots", "rate0", "rate1", "t0",
              "t1" ; list(emp) :: "size" :: "initialRate0" :: "initialRate1"
              :: now :: now ) ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null,
              Code: call( "label:0" ; "this" ; "run0" ; emp ) ;
              free( "label:0" ) ; call( "label:1" ; "this" ; "run1" ; emp ) ;
              free( "label:1" ) ; return ( emp ) >,
            < "run0" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null,
              Code: call( "label:0" ; "this" ; "read0" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ; await ">=" ( now
              :: "+" ( "t0" :: "rate0" ) ) ; posit "<=" ( now :: "+" ( "t0"
              :: "rate0" ) ) ; assign( "t0" ; now ) ;
              call( "label:1" ; "this" ; "run0" ; emp ) ; free( "label:1" ) ;
              return ( emp ) >,
            < "read0" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "temp" |-> null, "_" |-> null,
              Code: call( "label:0" ; "sensor0" ; "read" ; emp ) ;
              get( "label:0" ; "temp" ) ; free( "label:0" ) ; posit
              "<=" ( now :: "t0" ) ; await ">" ( "slots" :: int(0) ) ;
              assign( "buffer" ; "|-" ( "buffer" :: "temp" ) ) ;
              assign( "slots" ; "-" ( "slots" :: int(1) ) ) ; if
              "=" ( "slots" :: int(0) ) th
              call( "label:1" ; "this" ; "packAndSend" ; emp ) ;
              get( "label:1" ; noVid ) ; free( "label:1" ) el skip fi ;
              return ( emp ) >,
            < "run1" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null,
              Code: call( "label:0" ; "this" ; "read1" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ; await ">=" ( now
              :: "+" ( "t1" :: "rate1" ) ) ; posit "<=" ( now :: "+" ( "t1"
              :: "rate1" ) ) ; assign( "t1" ; now ) ;
              call( "label:1" ; "this" ; "run1" ; emp ) ; free( "label:1" ) ;
              return ( emp ) >,
            < "read1" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "temp" |-> null, "_" |-> null,
              Code: call( "label:0" ; "sensor1" ; "read" ; emp ) ;
              get( "label:0" ; "temp" ) ; free( "label:0" ) ; posit
              "<=" ( now :: "t1" ) ; await ">" ( "slots" :: int(0) ) ;
              assign( "buffer" ; "|-" ( "buffer" :: "temp" ) ) ;
              assign( "slots" ; "-" ( "slots" :: int(1) ) ) ; if
              "=" ( "slots" :: int(0) ) th
              call( "label:1" ; "this" ; "packAndSend" ; emp ) ;
              get( "label:1" ; noVid ) ; free( "label:1" ) el skip fi ;
              return ( emp ) >,
            < "packAndSend" : Method | Param: noVid, Att: "label:0" |-> null,
              "_" |-> null, Code: call( "label:0" ; "radio" ; "write" ;
              "-|" ( "this" :: "-|" ( now :: "buffer" ) ) ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ;
              assign( "buffer" ; list(emp) ) ; assign( "slots" ; "size" ) ;
              return ( emp ) >,
            < "setRate" : Method | Param: "newRate", Att: "newRate" |-> null,
              "label:0" |-> null, "_" |-> null, Code: if
              "&&" ( "<" ( "newRate" :: "rate0" ) :: ">" ( now :: "+" ( "t0"
              :: "newRate" ) ) ) th assign( "t0" ; now ) ;
              call( "label:0" ; "this" ; "read0" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) el skip fi ;
              assign( "rate0" ; "newRate" ) ; posit "<" ( now :: "+" ( "t0"
              :: "rate0" ) ) ; return ( emp ) >,
            < "write" : Method | Param: "data", Att: "data" |-> null, "_" |->
              null, Code: skip ; return ( emp ) >,
      Ocnt: 0 >
    
    < "ActiveSensor" : Class | Inh: noInh, Param: "controller",
      "initialRate", Att: "controller" |-> null, "initialRate" |-> null,
      "rate" |-> null, "t0" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "rate", "t0" ; "initialRate" :: now ) ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:0" |-> null, "_" |->
              null, Code: await ">=" ( now :: "+" ( "t0" :: "rate" ) ) ;
              call( "label:0" ; "controller" ; "write" ; int(37) ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ; posit "<=" ( now
              :: "+" ( "t0" :: "rate" ) ) ; assign( "t0" ; now ) ;
              return ( emp ) >,
            < "setRate" : Method | Param: "newRate", Att: "newRate" |-> null,
              "label:0" |-> null, "_" |-> null, Code: if
              "&&" ( "<" ( "newRate" :: "rate" ) :: ">" ( now :: "+" ( "t0"
              :: "newRate" ) ) ) th call( "label:0" ; "controller" ;
              "write" ; int(37) ) ; get( "label:0" ; noVid ) ;
              free( "label:0" ) ; assign( "t0" ; now ) el skip fi ;
              assign( "rate" ; "newRate" ) ; posit "<" ( now :: "+" ( "t0" ::
              "rate" ) ) ; return ( emp ) >,
            < "setResolution" : Method | Param: "data", Att: "data" |-> null,
              "_" |-> null, Code: skip ; return ( emp ) >,
            < "setEncoding" : Method | Param: "data", Att: "data" |-> null,
              "_" |-> null, Code: skip ; return ( emp ) >,
            < "switchOn" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
            < "switchOff" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
      Ocnt: 0 > .
endm

