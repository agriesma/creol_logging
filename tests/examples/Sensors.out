==========================================
reduce in PROGRAM : state .
result State: {
< "ActiveSensor" : Class | Inh: noInh, Param: "controller", "initialRate", Att:
    "controller" |-> null, "initialRate" |-> null, "rate" |-> null, "t0" |->
    null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("rate",
    "t0" ; "initialRate" :: now) ; return(emp) >, < "run" : Method | Param:
    noVid, Att: "_" |-> null, "ccglab0" |-> null, Code: await ">="(now :: "+"(
    "t0" :: "rate")) ; call("ccglab0" ; "controller" ; "write" ; int(37)) ;
    get("ccglab0" ; noVid) ; free("ccglab0") ; posit "<="(now :: "+"("t0" ::
    "rate")) ; assign("t0" ; now) ; return(emp) >, < "setEncoding" : Method |
    Param: "data", Att: "_" |-> null, "data" |-> null, Code: skip ; return(emp)
    >, < "setRate" : Method | Param: "newRate", Att: "_" |-> null, "ccglab0"
    |-> null, "newRate" |-> null, Code: if "&&"("<"("newRate" :: "rate") ::
    ">"(now :: "+"("t0" :: "newRate"))) th call("ccglab0" ; "controller" ;
    "write" ; int(37)) ; get("ccglab0" ; noVid) ; free("ccglab0") ; assign("t0"
    ; now) el skip fi ; assign("rate" ; "newRate") ; posit "<"(now :: "+"("t0"
    :: "rate")) ; return(emp) >, < "setResolution" : Method | Param: "data",
    Att: "_" |-> null, "data" |-> null, Code: skip ; return(emp) >, <
    "switchOff" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
    return(emp) >, < "switchOn" : Method | Param: noVid, Att: "_" |-> null,
    Code: skip ; return(emp) >,
  Ocnt: 0 >
 
< "PassiveTempSensor" : Class | Inh: noInh, Param: noVid, Att: noSubst, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "read" : Method | Param: noVid, Att: "_" |-> null, "value" |-> null,
    Code: assign("value" ; int(37)) ; return("value") >, < "run" : Method |
    Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >, <
    "setEncoding" : Method | Param: "encoding", Att: "_" |-> null, "encoding"
    |-> null, Code: skip ; return(emp) >, < "setRate" : Method | Param:
    "newRate", Att: "_" |-> null, "newRate" |-> null, Code: skip ; return(emp)
    >, < "setResolution" : Method | Param: "res", Att: "_" |-> null, "res" |->
    null, Code: skip ; return(emp) >, < "switchOff" : Method | Param: noVid,
    Att: "_" |-> null, Code: skip ; return(emp) >, < "switchOn" : Method |
    Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >
 
< "SimpleSensorNode" : Class | Inh: noInh, Param: "sensor", "radio", "size",
    "initialRate", Att: "buffer" |-> null, "initialRate" |-> null, "radio" |->
    null, "rate" |-> null, "sensor" |-> null, "size" |-> null, "slots" |->
    null, "t0" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("buffer" ;
    list(emp)) ; assign("slots" ; "size") ; assign("rate" ; "initialRate") ;
    assign("t0" ; now) ; return(emp) >, < "packAndSend" : Method | Param:
    noVid, Att: "_" |-> null, "ccglab0" |-> null, Code: call("ccglab0" ;
    "radio" ; "write" ; "-|"("this" :: "-|"(now :: "buffer"))) ; get("ccglab0"
    ; noVid) ; free("ccglab0") ; assign("buffer" ; list(emp)) ; assign("slots"
    ; "size") ; return(emp) >, < "read" : Method | Param: noVid, Att: "_" |->
    null, "ccglab0" |-> null, "ccglab1" |-> null, "temp" |-> null, Code: call(
    "ccglab0" ; "sensor" ; "read" ; emp) ; get("ccglab0" ; "temp") ; free(
    "ccglab0") ; posit "<="(now :: "t0") ; await ">"("slots" :: int(0)) ;
    assign("buffer" ; "|-"("buffer" :: "temp")) ; assign("slots" ; "-"("slots"
    :: int(1))) ; if "="("slots" :: int(0)) th call("ccglab1" ; "this" ;
    "packAndSend" ; emp) ; get("ccglab1" ; noVid) ; free("ccglab1") el skip fi
    ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null,
    "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0" ; "this" ;
    "read" ; emp) ; get("ccglab0" ; noVid) ; free("ccglab0") ; await ">="(now
    :: "+"("t0" :: "rate")) ; posit "<="(now :: "+"("t0" :: "rate")) ; assign(
    "t0" ; now) ; call("ccglab1" ; "this" ; "run" ; emp) ; free("ccglab1") ;
    return(emp) >, < "setRate" : Method | Param: "newRate", Att: "_" |-> null,
    "ccglab0" |-> null, "newRate" |-> null, Code: if "&&"("<"("newRate" ::
    "rate") :: ">"(now :: "+"("t0" :: "newRate"))) th assign("t0" ; now) ;
    call("ccglab0" ; "this" ; "read" ; emp) ; get("ccglab0" ; noVid) ; free(
    "ccglab0") el skip fi ; assign("rate" ; "newRate") ; posit "<"(now :: "+"(
    "t0" :: "rate")) ; return(emp) >, < "write" : Method | Param: "data", Att:
    "_" |-> null, "data" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >
 
< "TwoSimpleSensorNode" : Class | Inh: noInh, Param: "sensor0", "sensor1",
    "radio", "size", "initialRate0", "initialRate1", Att: "buffer" |-> null,
    "initialRate0" |-> null, "initialRate1" |-> null, "radio" |-> null, "rate0"
    |-> null, "rate1" |-> null, "sensor0" |-> null, "sensor1" |-> null, "size"
    |-> null, "slots" |-> null, "t0" |-> null, "t1" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("buffer",
    "slots", "rate0", "rate1", "t0", "t1" ; list(emp) :: "size" ::
    "initialRate0" :: "initialRate1" :: now :: now) ; return(emp) >, <
    "packAndSend" : Method | Param: noVid, Att: "_" |-> null, "ccglab0" |->
    null, Code: call("ccglab0" ; "radio" ; "write" ; "-|"("this" :: "-|"(now ::
    "buffer"))) ; get("ccglab0" ; noVid) ; free("ccglab0") ; assign("buffer" ;
    list(emp)) ; assign("slots" ; "size") ; return(emp) >, < "read0" : Method |
    Param: noVid, Att: "_" |-> null, "ccglab0" |-> null, "ccglab1" |-> null,
    "temp" |-> null, Code: call("ccglab0" ; "sensor0" ; "read" ; emp) ; get(
    "ccglab0" ; "temp") ; free("ccglab0") ; posit "<="(now :: "t0") ; await
    ">"("slots" :: int(0)) ; assign("buffer" ; "|-"("buffer" :: "temp")) ;
    assign("slots" ; "-"("slots" :: int(1))) ; if "="("slots" :: int(0)) th
    call("ccglab1" ; "this" ; "packAndSend" ; emp) ; get("ccglab1" ; noVid) ;
    free("ccglab1") el skip fi ; return(emp) >, < "read1" : Method | Param:
    noVid, Att: "_" |-> null, "ccglab0" |-> null, "ccglab1" |-> null, "temp"
    |-> null, Code: call("ccglab0" ; "sensor1" ; "read" ; emp) ; get("ccglab0"
    ; "temp") ; free("ccglab0") ; posit "<="(now :: "t1") ; await ">"("slots"
    :: int(0)) ; assign("buffer" ; "|-"("buffer" :: "temp")) ; assign("slots" ;
    "-"("slots" :: int(1))) ; if "="("slots" :: int(0)) th call("ccglab1" ;
    "this" ; "packAndSend" ; emp) ; get("ccglab1" ; noVid) ; free("ccglab1") el
    skip fi ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |->
    null, "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0" ; "this"
    ; "run0" ; emp) ; free("ccglab0") ; call("ccglab1" ; "this" ; "run1" ; emp)
    ; free("ccglab1") ; return(emp) >, < "run0" : Method | Param: noVid, Att:
    "_" |-> null, "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0"
    ; "this" ; "read0" ; emp) ; get("ccglab0" ; noVid) ; free("ccglab0") ;
    await ">="(now :: "+"("t0" :: "rate0")) ; posit "<="(now :: "+"("t0" ::
    "rate0")) ; assign("t0" ; now) ; call("ccglab1" ; "this" ; "run0" ; emp) ;
    free("ccglab1") ; return(emp) >, < "run1" : Method | Param: noVid, Att: "_"
    |-> null, "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0" ;
    "this" ; "read1" ; emp) ; get("ccglab0" ; noVid) ; free("ccglab0") ; await
    ">="(now :: "+"("t1" :: "rate1")) ; posit "<="(now :: "+"("t1" :: "rate1"))
    ; assign("t1" ; now) ; call("ccglab1" ; "this" ; "run1" ; emp) ; free(
    "ccglab1") ; return(emp) >, < "setRate" : Method | Param: "newRate", Att:
    "_" |-> null, "ccglab0" |-> null, "newRate" |-> null, Code: if "&&"("<"(
    "newRate" :: "rate0") :: ">"(now :: "+"("t0" :: "newRate"))) th assign("t0"
    ; now) ; call("ccglab0" ; "this" ; "read0" ; emp) ; get("ccglab0" ; noVid)
    ; free("ccglab0") el skip fi ; assign("rate0" ; "newRate") ; posit "<"(now
    :: "+"("t0" :: "rate0")) ; return(emp) >, < "write" : Method | Param:
    "data", Att: "_" |-> null, "data" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >
}
Bye.
